### PR TITLE
fix crash after identify and alias

### PR DESCRIPTION
**Pull request by aboedo**
Fixes a bug in #232 that would cause identify and alias to crash. 

Also, fixes race conditions that we've had here forever, since you could potentially have a nil appUserID between clearing cache and saving a new appUserID

Missing tests but reviewable otherwise

### CODE REVIEW

File: Purchases/RCIdentityManager.m [hunk 1]
```diff
@@ -51,51 +49,43 @@ - (void)configureWithAppUserID:(nullable NSString *)appUserID
...
         completion(error);
     }];
 }
 
-- (void)resetAppUserID
-{
-    [self.deviceCache clearCachesForAppUserID:self.currentAppUserID];
+- (void)resetAppUserID {
     NSString *randomId = [self generateRandomID];
     [self saveAppUserID:randomId];
```
Comment (vegaro) [inline Purchases/RCIdentityManager.m]: I think this is not needed

File: Purchases/Caching/RCDeviceCache.m [hunk 2]
```diff
@@ -101,15 +101,16 @@ - (void)cacheAppUserID:(NSString *)appUserID {
...
-        [self.userDefaults removeObjectForKey:RCAppUserDefaultsKey];
-        [self.userDefaults removeObjectForKey:[self purchaserInfoUserDefaultCacheKeyForAppUserID:appUserID]];
+        [self.userDefaults removeObjectForKey:[self purchaserInfoUserDefaultCacheKeyForAppUserID:oldAppUserID]];
         [self clearPurchaserInfoCacheTimestamp];
         [self clearOfferingsCache];
 
-        [self deleteAttributesIfSyncedForAppUserID:appUserID];
+        [self deleteAttributesIfSyncedForAppUserID:oldAppUserID];
+
+        [self.userDefaults setObject:newUserID forKey:RCAppUserDefaultsKey];
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:113]: There's a function `- (void)cacheAppUserID:(NSString *)appUserID {` in this class that already does this, so this can be replaced with `[self cacheAppUserID:newUserID]`

File: Purchases/RCIdentityManager.m (hunk 1 above)
Comment (aboedo) [reply Purchases/RCIdentityManager.m]: good catch, I'd missed that

File: Purchases/Caching/RCDeviceCache.m (hunk 2 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:113]: yep, updating

**Comment by aboedo** [comment]
added tests and implemented changes from comments, @vegaro ready for another pass

File: Purchases/Caching/RCDeviceCache.m [hunk 3]
```diff
@@ -101,15 +101,16 @@ - (void)cacheAppUserID:(NSString *)appUserID {
     [self.userDefaults setObject:appUserID forKey:RCAppUserDefaultsKey];
 }
 
-- (void)clearCachesForAppUserID:(NSString *)appUserID {
+- (void)clearCachesForAppUserID:(NSString *)oldAppUserID andSaveNewUserID:(NSString *)newUserID {
     @synchronized (self) {
         [self.userDefaults removeObjectForKey:RCLegacyGeneratedAppUserDefaultsKey];
-        [self.userDefaults removeObjectForKey:RCAppUserDefaultsKey];
```
Comment (vegaro) [inline Purchases/Caching/RCDeviceCache.m:107]: Why was this removed?

File: PurchasesTests/Caching/DeviceCacheTests.swift [hunk 4]
```diff
@@ -37,63 +37,63 @@ class DeviceCacheTests: XCTestCase {
...
 
-    func testClearCachesClearsCachesTimestamp() {
+    func testClearCachesForAppUserIDAndSaveNewUserIDClearsCachesTimestamp() {
         self.deviceCache.setPurchaserInfoCacheTimestampToNow()
-        self.deviceCache.clearCaches(forAppUserID: "cesar")
+        self.deviceCache.clearCaches(forAppUserID: "cesar", andSaveNewUserID: "newUser")
         expect(self.deviceCache.isPurchaserInfoCacheStale()).to(beTrue())
     }
 
-    func testClearCachesRemovesCachedAppUserIDs() {
```
Comment (vegaro) [inline PurchasesTests/Caching/DeviceCacheTests.swift:60]: This test name wasn't very descriptive, but it was testing the migration to the newer `appUserIDs` we added on Identity 2.0. I think we should keep it since the migration can still be needed.

File: Purchases/Caching/RCDeviceCache.m (hunk 3 above)
Comment (aboedo) [reply Purchases/Caching/RCDeviceCache.m:107]: it was replaced by `self.cacheAppUserID`, so you never have a nil appUserID, which is an invalid state

File: PurchasesTests/Caching/DeviceCacheTests.swift (hunk 4 above)
Comment (aboedo) [reply PurchasesTests/Caching/DeviceCacheTests.swift:60]: it's not really removed, it's replaced with `testClearCachesForAppUserIDAndSaveNewUserIDUpdatesCachedAppUserID`. 
however, there's a missing line here that should be checking that we remove the legacy appUserID, I'll add it

**Comment by aboedo** [comment]
@vegaro obviously android doesn't have the same crash, but the race condition in IdentityManager where we delete the app user id and store a new one as separate operations is there as well I believe, we may wanna either synchronize the calls or wrap them into a single one like in this PR
